benchmarking/compute_probability_of_improvement.py

```diff
@@ -0,0 +1,251 @@
+"""Compute the probability that one algorithm improved over another."""
+import argparse
+import pathlib
+import sys
+import warnings
+from typing import Dict, List, Optional
+
+import numpy as np
+from rliable import library as rly
+from rliable import metrics
+
+from imitation.util.sacred_file_parsing import SacredRun, group_runs_by_algo_and_env
+
+
+def sample_matrix_from_runs_by_env(
+    runs_by_env: Dict[str, List[SacredRun]],
+    envs: Optional[List[str]] = None,
+) -> np.ndarray:
+    """Samples a matrix of scores from the runs for each environment.
+
+    Note: when the number of samples for each environment is not equal, the samples
+    will be truncated to the minimum sample count.
+
+    Args:
+        runs_by_env: A dictionary mapping environment names to lists of runs.
+        envs: The environments to sample from. If None, all environments are used.
+
+    Returns:
+        A matrix of scores of shape (n_samples, n_envs).
+    """
+    if envs is None:
+        envs = list(runs_by_env.keys())
+
+    sample_counts_by_env = {env: len(runs_by_env[env]) for env in envs}
+
+    min_sample_count = min(sample_counts_by_env.values())
+    if not all(
+        sample_counts_by_env[env] == sample_counts_by_env[envs[0]] for env in envs
+    ):
+        warnings.warn(
+            f"The runs for the environments have different sample counts "
+            f"{sample_counts_by_env}. "
+            f"This is not supported by the probability of improvement. Therefore, "
+            f"samples will be truncated to the minimum sample count of"
+            f" {min_sample_count}",
+        )
+
+    return np.asarray(
+        [
+            [
+                run["result"]["imit_stats"]["monitor_return_mean"]
+                for run in runs_by_env[env][:min_sample_count]
+            ]
+            for env in envs
+        ],
+    ).T
+
+
+def compute_probability_of_improvement(
+    runs_by_env: Dict[str, List[SacredRun]],
+    baseline_runs_by_env: Dict[str, List[SacredRun]],
+    reps: int,
+):
+    """Computes the probability of improvement of the runs over the baseline runs.
+
+    Args:
+        runs_by_env: A dictionary mapping environment names to lists of runs.
+        baseline_runs_by_env: A dictionary mapping environment names to lists of runs.
+        reps: The number of bootstrap repetitions to use to compute the confidence
+            interval.
+
+    Returns:
+        A tuple of:
+         - probability of improvement
+         - confidence interval
+         - number of samples per env
+         - number of baseline samples per env
+    """
+    envs = runs_by_env.keys()
+    baseline_envs = baseline_runs_by_env.keys()
+    comparison_envs = sorted(set(envs).intersection(set(baseline_envs)))
+
+    run_scores = sample_matrix_from_runs_by_env(runs_by_env, comparison_envs)
+    baseline_run_scores = sample_matrix_from_runs_by_env(
+        baseline_runs_by_env,
+        comparison_envs,
+    )
```

Comment:
Suggestion: add comment explaining dimensions & an assertion that the shape makes sense (in particular `run_scores.shape` should match the pattern `(s, len(comparison_envs))`, where `s>=1`).

Reply:
Yes, this part of the code is a bit confusing. Added some asserts and clarification.